### ISSUE TITLE
Provenance Design

### ISSUE THREAD

**Issue by benloh** [netcreateorg/netcreate-itest#37]
See https://docs.google.com/document/d/174JMW5ktUbS3Chygu82dwMdtUlkSk6ef4XcWwnFxXSE/edit

**Comment by benloh** [comment]
# Provenance Design WIP

Provenance fields need to be defined on a per-project basis.  Each project might have its own set of fields.

Provenance is the same for both nodes and edges.

Can you select a provenance type ala comments?
37

## Features

#### Default (Simple) Provenance

```
provenanceDefs = [
  { field: 'Source', type: 'text'}
]
```

#### Standard Provenance
```
provenanceDefs = [
  { field: 'Author', type: 'text'}
  { field: 'Date', type: 'timestamp'}
  { field: 'Notes', type: 'text'}
]
```


## Data & Templating

Use the comment prompt template approach of stuffing provenance into a single string that is dereferenced and rendered based on the provenance template specification.

The advantage to this approach is that all of the provenance is captured as a single string, rather than a series of separate fields that have to be separate managed into the provenance tab.  For example, we don't need to add an extra field designating a custom field as a provenance field.

Defining the template is also easier, e.g. if a teacher is doing this they don't have to make decisions about displayLabels vs exportLabels, hidden, etc.

Defining the provenance fields is done via a separate simplified template editor, ala Node and Edge labels/colors.  

## UI





Node schema
```
node = {
  ...
  provenance: [
     {label: 'Author', type: 'text'},
     {label: 'Date', type: 'date-year'),
     {label: 'Notes', 'type: 'text')
  ],
  created: {author: 'Joshua', utc: 1688663266872), // system time in utc?
  updated: {author: 'Joshua', utc: 1688663266872),
  ...
} 
```

**Comment by benloh** [comment]
@jdanish Just verifying that this is what you want.  If we use "Source" as the display name, Edges would look like this:

![screenshot_1573](https://github.com/netcreateorg/netcreate-itest/assets/1403057/1acc29a5-eea0-4453-8091-427384a1f224)

If that's not a problem we can display it that way and still use `infoSource` as the id, and `InfoSource` as the table column heading for exports.

We can also use "Info Source" as the display name in the edge of course.

**Comment by jdanish** [comment]
Let's use **Info Source** for the label for now. I assume we'll be able to edit it in the template?  Otherwise, yes.

I'm not sure I follow the logic on treating Provenance fields as "special" ... my hope is that the template system will be intuitive, teachers will have to use it anyhow for nodes / edges so need to figure out those details anyhow (and thus using a template and sub-template is confusing), and we can provide defaults (e.g., visible, same label and export, etc). 

Also, in case we were on different pages - I was assuming we'd reduce the mandatory fields to the minimum (label, id, target, source?) and everything else would be flexible and arbitrary?  That's why the table needs to be more flexible / modifiable.

**Comment by jdanish** [comment]
Also, I hope no one ever names a node source or sourcetarget ... it's painful.

**Comment by benloh** [comment]
> Let's use Info Source for the label for now. I assume we'll be able to edit it in the template? Otherwise, yes.

Yes.  The display label is editable in the template.

> I'm not sure I follow the logic on treating Provenance fields as "special" ... my hope is that the template system will be intuitive, teachers will have to use it anyhow for nodes / edges so need to figure out those details anyhow (and thus using a template and sub-template is confusing), and we can provide defaults (e.g., visible, same label and export, etc).

Sorry, those were old notes that I hadn't finished removing.  I wanted to keep those before updating the other comments.

> Also, I hope no one ever names a node source or sourcetarget ... it's painful.

Heh.  I had too many fields in my test project and needed to keep track of it.

**Comment by jdanish** [comment]
Cool. Then yes sounds good to me.

**Comment by benloh** [comment]
@jdanish Here's a quick semi-working mockup.  Right idea?
![screenshot_1574](https://github.com/netcreateorg/netcreate-itest/assets/1403057/0b7d1e90-b996-4f8a-ad24-7dd77108737a)

**Comment by benloh** [comment]
@jdanish One more question: Should Node Tables and Edge Tables show all custom provenance fields for every row?  It does start to get to be a LOT of data for a table row.  Just show it with a popup on hover?

**Comment by jdanish** [comment]
Yes that looks good.

For the other question … I think ideally we’d want some template options for all columns and even better if kids can change some on the fly…. Meaning some networks we might want one field or the other or want kids to decide. Others none. Others maybe a summary would work (special case for provenance but not for other attributes).

**Comment by benloh** [comment]
>> @jdanish One more question: Should Node Tables and Edge Tables show all custom provenance fields for every row? It does start to get to be a LOT of data for a table row. Just show it with a popup on hover?

> For the other question … I think ideally we’d want some template options for all columns and even better if kids can change some on the fly…. Meaning some networks we might want one field or the other or want kids to decide. Others none. Others maybe a summary would work (special case for provenance but not for other attributes).

Do you mean you want the ability to show/hide table provenance individual table columns?  I'm a little worried about things getting unnecessarily complex (not that it's hard to do, but have to add a UI to manage it).  There's already so many parameters that you have to tweak on and off I worry that managing templates will be a nightmare.

Note also that all of these changes represent a major change to the data format, so we'll either have to migrate old versions of project files, or make everyone create new ones.  I suspect we'll want a graceful migration.

**Comment by jdanish** [comment]
Yeah... I share your concerns...  

Re: Template: I think we'll want to think about how to use "normal" and "advanced" properties  similar to the GEM-STEP code panels to help with that (And then defaults on all of the advanced things to simplify it), but it's still a concern.

Yes, graceful migration is the ideal.

[Edited.  Moved to #108]

**Comment by jdanish** [comment]
[Edited.  Moved to #108]

**Comment by benloh** [comment]
I hear you about the table issues.  But we already have a number of issues logged to deal with that.  So I'm gonna edit the comments and move them to the table issue #108.

**Comment by benloh** [comment]
Implemented with #213
